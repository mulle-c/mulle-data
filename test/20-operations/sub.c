#include <mulle-data/mulle-data.h>

#include <stdio.h>

static void  mulle_range_print( struct mulle_range a)
{
   if( ! a.length)
      printf( "NULL");  
   else
      printf( "%td-%td (%td)", a.location, a.location + a.length - 1, a.length);  
}


int   main( int argc, char *argv[])
{
   struct mulle_range   a;
   struct mulle_range   b;
   struct mulle_range   c[ 2];
   unsigned int         i, n;

   for( a.location = 0; a.location < 100; a.location += 25)
      for( b.location = 0; b.location < 100; b.location += 25)
         for( a.length = 0; a.length < 100; a.length += 25)
            for( b.length = 0; b.length < 100; b.length += 25)
//   a = mulle_range_make( 25, 25);
//   b = mulle_range_make( 0, 75);

            {
               n = mulle_range_subtract( a, b, c);
               mulle_range_print( a);
               printf( " - ");
               mulle_range_print( b);
               printf( " = ");
               for( i = 0; i < n; i++)
               {
                  assert( mulle_range_is_valid( c[ i]));
                  if( i)
                     printf( ", ");
                  mulle_range_print( c[ i]); 
               }
               printf( "\n");
            }

   return( 0);
}


/*
 * extension : mulle-sde/c-test-library-demo
 * directory : demo/all
 * template  : .../hello.c
 * Suppress this comment with `export MULLE_SDE_GENERATE_FILE_COMMENTS=NO`
 */
