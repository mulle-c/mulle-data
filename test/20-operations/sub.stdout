NULL - NULL = NULL
NULL - 0-24 (25) = NULL
NULL - 0-49 (50) = NULL
NULL - 0-74 (75) = NULL
0-24 (25) - NULL = 0-24 (25)
0-24 (25) - 0-24 (25) = NULL
0-24 (25) - 0-49 (50) = NULL
0-24 (25) - 0-74 (75) = NULL
0-49 (50) - NULL = 0-49 (50)
0-49 (50) - 0-24 (25) = 25-49 (25)
0-49 (50) - 0-49 (50) = NULL
0-49 (50) - 0-74 (75) = NULL
0-74 (75) - NULL = 0-74 (75)
0-74 (75) - 0-24 (25) = 25-74 (50)
0-74 (75) - 0-49 (50) = 50-74 (25)
0-74 (75) - 0-74 (75) = NULL
NULL - NULL = NULL
NULL - 25-49 (25) = NULL
NULL - 25-74 (50) = NULL
NULL - 25-99 (75) = NULL
0-24 (25) - NULL = 0-24 (25)
0-24 (25) - 25-49 (25) = 0-24 (25)
0-24 (25) - 25-74 (50) = 0-24 (25)
0-24 (25) - 25-99 (75) = 0-24 (25)
0-49 (50) - NULL = 0-49 (50)
0-49 (50) - 25-49 (25) = 0-24 (25)
0-49 (50) - 25-74 (50) = 0-24 (25)
0-49 (50) - 25-99 (75) = 0-24 (25)
0-74 (75) - NULL = 0-74 (75)
0-74 (75) - 25-49 (25) = 0-24 (25), 50-74 (25)
0-74 (75) - 25-74 (50) = 0-24 (25)
0-74 (75) - 25-99 (75) = 0-24 (25)
NULL - NULL = NULL
NULL - 50-74 (25) = NULL
NULL - 50-99 (50) = NULL
NULL - 50-124 (75) = NULL
0-24 (25) - NULL = 0-24 (25)
0-24 (25) - 50-74 (25) = 0-24 (25)
0-24 (25) - 50-99 (50) = 0-24 (25)
0-24 (25) - 50-124 (75) = 0-24 (25)
0-49 (50) - NULL = 0-49 (50)
0-49 (50) - 50-74 (25) = 0-49 (50)
0-49 (50) - 50-99 (50) = 0-49 (50)
0-49 (50) - 50-124 (75) = 0-49 (50)
0-74 (75) - NULL = 0-74 (75)
0-74 (75) - 50-74 (25) = 0-49 (50)
0-74 (75) - 50-99 (50) = 0-49 (50)
0-74 (75) - 50-124 (75) = 0-49 (50)
NULL - NULL = NULL
NULL - 75-99 (25) = NULL
NULL - 75-124 (50) = NULL
NULL - 75-149 (75) = NULL
0-24 (25) - NULL = 0-24 (25)
0-24 (25) - 75-99 (25) = 0-24 (25)
0-24 (25) - 75-124 (50) = 0-24 (25)
0-24 (25) - 75-149 (75) = 0-24 (25)
0-49 (50) - NULL = 0-49 (50)
0-49 (50) - 75-99 (25) = 0-49 (50)
0-49 (50) - 75-124 (50) = 0-49 (50)
0-49 (50) - 75-149 (75) = 0-49 (50)
0-74 (75) - NULL = 0-74 (75)
0-74 (75) - 75-99 (25) = 0-74 (75)
0-74 (75) - 75-124 (50) = 0-74 (75)
0-74 (75) - 75-149 (75) = 0-74 (75)
NULL - NULL = NULL
NULL - 0-24 (25) = NULL
NULL - 0-49 (50) = NULL
NULL - 0-74 (75) = NULL
25-49 (25) - NULL = 25-49 (25)
25-49 (25) - 0-24 (25) = 25-49 (25)
25-49 (25) - 0-49 (50) = NULL
25-49 (25) - 0-74 (75) = NULL
25-74 (50) - NULL = 25-74 (50)
25-74 (50) - 0-24 (25) = 25-74 (50)
25-74 (50) - 0-49 (50) = 50-74 (25)
25-74 (50) - 0-74 (75) = NULL
25-99 (75) - NULL = 25-99 (75)
25-99 (75) - 0-24 (25) = 25-99 (75)
25-99 (75) - 0-49 (50) = 50-99 (50)
25-99 (75) - 0-74 (75) = 75-99 (25)
NULL - NULL = NULL
NULL - 25-49 (25) = NULL
NULL - 25-74 (50) = NULL
NULL - 25-99 (75) = NULL
25-49 (25) - NULL = 25-49 (25)
25-49 (25) - 25-49 (25) = NULL
25-49 (25) - 25-74 (50) = NULL
25-49 (25) - 25-99 (75) = NULL
25-74 (50) - NULL = 25-74 (50)
25-74 (50) - 25-49 (25) = 50-74 (25)
25-74 (50) - 25-74 (50) = NULL
25-74 (50) - 25-99 (75) = NULL
25-99 (75) - NULL = 25-99 (75)
25-99 (75) - 25-49 (25) = 50-99 (50)
25-99 (75) - 25-74 (50) = 75-99 (25)
25-99 (75) - 25-99 (75) = NULL
NULL - NULL = NULL
NULL - 50-74 (25) = NULL
NULL - 50-99 (50) = NULL
NULL - 50-124 (75) = NULL
25-49 (25) - NULL = 25-49 (25)
25-49 (25) - 50-74 (25) = 25-49 (25)
25-49 (25) - 50-99 (50) = 25-49 (25)
25-49 (25) - 50-124 (75) = 25-49 (25)
25-74 (50) - NULL = 25-74 (50)
25-74 (50) - 50-74 (25) = 25-49 (25)
25-74 (50) - 50-99 (50) = 25-49 (25)
25-74 (50) - 50-124 (75) = 25-49 (25)
25-99 (75) - NULL = 25-99 (75)
25-99 (75) - 50-74 (25) = 25-49 (25), 75-99 (25)
25-99 (75) - 50-99 (50) = 25-49 (25)
25-99 (75) - 50-124 (75) = 25-49 (25)
NULL - NULL = NULL
NULL - 75-99 (25) = NULL
NULL - 75-124 (50) = NULL
NULL - 75-149 (75) = NULL
25-49 (25) - NULL = 25-49 (25)
25-49 (25) - 75-99 (25) = 25-49 (25)
25-49 (25) - 75-124 (50) = 25-49 (25)
25-49 (25) - 75-149 (75) = 25-49 (25)
25-74 (50) - NULL = 25-74 (50)
25-74 (50) - 75-99 (25) = 25-74 (50)
25-74 (50) - 75-124 (50) = 25-74 (50)
25-74 (50) - 75-149 (75) = 25-74 (50)
25-99 (75) - NULL = 25-99 (75)
25-99 (75) - 75-99 (25) = 25-74 (50)
25-99 (75) - 75-124 (50) = 25-74 (50)
25-99 (75) - 75-149 (75) = 25-74 (50)
NULL - NULL = NULL
NULL - 0-24 (25) = NULL
NULL - 0-49 (50) = NULL
NULL - 0-74 (75) = NULL
50-74 (25) - NULL = 50-74 (25)
50-74 (25) - 0-24 (25) = 50-74 (25)
50-74 (25) - 0-49 (50) = 50-74 (25)
50-74 (25) - 0-74 (75) = NULL
50-99 (50) - NULL = 50-99 (50)
50-99 (50) - 0-24 (25) = 50-99 (50)
50-99 (50) - 0-49 (50) = 50-99 (50)
50-99 (50) - 0-74 (75) = 75-99 (25)
50-124 (75) - NULL = 50-124 (75)
50-124 (75) - 0-24 (25) = 50-124 (75)
50-124 (75) - 0-49 (50) = 50-124 (75)
50-124 (75) - 0-74 (75) = 75-124 (50)
NULL - NULL = NULL
NULL - 25-49 (25) = NULL
NULL - 25-74 (50) = NULL
NULL - 25-99 (75) = NULL
50-74 (25) - NULL = 50-74 (25)
50-74 (25) - 25-49 (25) = 50-74 (25)
50-74 (25) - 25-74 (50) = NULL
50-74 (25) - 25-99 (75) = NULL
50-99 (50) - NULL = 50-99 (50)
50-99 (50) - 25-49 (25) = 50-99 (50)
50-99 (50) - 25-74 (50) = 75-99 (25)
50-99 (50) - 25-99 (75) = NULL
50-124 (75) - NULL = 50-124 (75)
50-124 (75) - 25-49 (25) = 50-124 (75)
50-124 (75) - 25-74 (50) = 75-124 (50)
50-124 (75) - 25-99 (75) = 100-124 (25)
NULL - NULL = NULL
NULL - 50-74 (25) = NULL
NULL - 50-99 (50) = NULL
NULL - 50-124 (75) = NULL
50-74 (25) - NULL = 50-74 (25)
50-74 (25) - 50-74 (25) = NULL
50-74 (25) - 50-99 (50) = NULL
50-74 (25) - 50-124 (75) = NULL
50-99 (50) - NULL = 50-99 (50)
50-99 (50) - 50-74 (25) = 75-99 (25)
50-99 (50) - 50-99 (50) = NULL
50-99 (50) - 50-124 (75) = NULL
50-124 (75) - NULL = 50-124 (75)
50-124 (75) - 50-74 (25) = 75-124 (50)
50-124 (75) - 50-99 (50) = 100-124 (25)
50-124 (75) - 50-124 (75) = NULL
NULL - NULL = NULL
NULL - 75-99 (25) = NULL
NULL - 75-124 (50) = NULL
NULL - 75-149 (75) = NULL
50-74 (25) - NULL = 50-74 (25)
50-74 (25) - 75-99 (25) = 50-74 (25)
50-74 (25) - 75-124 (50) = 50-74 (25)
50-74 (25) - 75-149 (75) = 50-74 (25)
50-99 (50) - NULL = 50-99 (50)
50-99 (50) - 75-99 (25) = 50-74 (25)
50-99 (50) - 75-124 (50) = 50-74 (25)
50-99 (50) - 75-149 (75) = 50-74 (25)
50-124 (75) - NULL = 50-124 (75)
50-124 (75) - 75-99 (25) = 50-74 (25), 100-124 (25)
50-124 (75) - 75-124 (50) = 50-74 (25)
50-124 (75) - 75-149 (75) = 50-74 (25)
NULL - NULL = NULL
NULL - 0-24 (25) = NULL
NULL - 0-49 (50) = NULL
NULL - 0-74 (75) = NULL
75-99 (25) - NULL = 75-99 (25)
75-99 (25) - 0-24 (25) = 75-99 (25)
75-99 (25) - 0-49 (50) = 75-99 (25)
75-99 (25) - 0-74 (75) = 75-99 (25)
75-124 (50) - NULL = 75-124 (50)
75-124 (50) - 0-24 (25) = 75-124 (50)
75-124 (50) - 0-49 (50) = 75-124 (50)
75-124 (50) - 0-74 (75) = 75-124 (50)
75-149 (75) - NULL = 75-149 (75)
75-149 (75) - 0-24 (25) = 75-149 (75)
75-149 (75) - 0-49 (50) = 75-149 (75)
75-149 (75) - 0-74 (75) = 75-149 (75)
NULL - NULL = NULL
NULL - 25-49 (25) = NULL
NULL - 25-74 (50) = NULL
NULL - 25-99 (75) = NULL
75-99 (25) - NULL = 75-99 (25)
75-99 (25) - 25-49 (25) = 75-99 (25)
75-99 (25) - 25-74 (50) = 75-99 (25)
75-99 (25) - 25-99 (75) = NULL
75-124 (50) - NULL = 75-124 (50)
75-124 (50) - 25-49 (25) = 75-124 (50)
75-124 (50) - 25-74 (50) = 75-124 (50)
75-124 (50) - 25-99 (75) = 100-124 (25)
75-149 (75) - NULL = 75-149 (75)
75-149 (75) - 25-49 (25) = 75-149 (75)
75-149 (75) - 25-74 (50) = 75-149 (75)
75-149 (75) - 25-99 (75) = 100-149 (50)
NULL - NULL = NULL
NULL - 50-74 (25) = NULL
NULL - 50-99 (50) = NULL
NULL - 50-124 (75) = NULL
75-99 (25) - NULL = 75-99 (25)
75-99 (25) - 50-74 (25) = 75-99 (25)
75-99 (25) - 50-99 (50) = NULL
75-99 (25) - 50-124 (75) = NULL
75-124 (50) - NULL = 75-124 (50)
75-124 (50) - 50-74 (25) = 75-124 (50)
75-124 (50) - 50-99 (50) = 100-124 (25)
75-124 (50) - 50-124 (75) = NULL
75-149 (75) - NULL = 75-149 (75)
75-149 (75) - 50-74 (25) = 75-149 (75)
75-149 (75) - 50-99 (50) = 100-149 (50)
75-149 (75) - 50-124 (75) = 125-149 (25)
NULL - NULL = NULL
NULL - 75-99 (25) = NULL
NULL - 75-124 (50) = NULL
NULL - 75-149 (75) = NULL
75-99 (25) - NULL = 75-99 (25)
75-99 (25) - 75-99 (25) = NULL
75-99 (25) - 75-124 (50) = NULL
75-99 (25) - 75-149 (75) = NULL
75-124 (50) - NULL = 75-124 (50)
75-124 (50) - 75-99 (25) = 100-124 (25)
75-124 (50) - 75-124 (50) = NULL
75-124 (50) - 75-149 (75) = NULL
75-149 (75) - NULL = 75-149 (75)
75-149 (75) - 75-99 (25) = 100-149 (50)
75-149 (75) - 75-124 (50) = 125-149 (25)
75-149 (75) - 75-149 (75) = NULL
